pkg/controllers/pod_controller.go

```diff
@@ -0,0 +1,290 @@
+			pod, ok := object.(*corev1.Pod)
+			return ok && usingExclusivePlacement(pod)
+		})).
+		Complete(r)
+}
+
+func SetupPodReconcilerIndexes(ctx context.Context, indexer client.FieldIndexer) error {
+	return indexer.IndexField(ctx, &corev1.Pod{}, podOwnerKey, func(obj client.Object) []string {
+		pod := obj.(*corev1.Pod)
+
+		// Make sure the pod is part of a JobSet using exclusive placement.
+		if _, exists := pod.Annotations[jobset.ExclusiveKey]; !exists {
+			return nil
+		}
+
+		// ...make sure it's a Pod owned by a Job...
+		owner := metav1.GetControllerOf(pod)
+		if owner == nil || owner.Kind != "Job" { // owner.APIVersion != batchv1.SchemeGroupVersion.String()
+			return nil
+		}
+		return []string{owner.Name}
+	})
+}
+
+// +kubebuilder:rbac:groups="",resources=events,verbs=create;watch;update;patch
+// +kubebuilder:rbac:groups=core,resources=pods,verbs=get;list;watch;create;update;patch;delete
+// +kubebuilder:rbac:groups=core,resources=nodes,verbs=get;list;watch
+
+// Reconcile is part of the main kubernetes reconciliation loop which aims to
+// move the current state of the cluster closer to the desired state.
+func (r *PodReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	log := ctrl.LoggerFrom(ctx)
+
+	// Get Pod from apiserver.
+	var pod corev1.Pod
+	if err := r.Get(ctx, req.NamespacedName, &pod); err != nil {
+		// we'll ignore not-found errors, since there is nothing we can do here.
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	// If user is not using exclusive placement, we don't need to do anything.
+	topologyKey, exists := pod.Annotations[jobset.ExclusiveKey]
+	if !exists {
+		return ctrl.Result{}, nil
+	}
+
+	// TODO: delete logging once we confirmed it works as it will be too noisy.
+	// log := ctrl.LoggerFrom(ctx).WithValues("pod", klog.KObj(&pod))
+	// ctx = ctrl.LoggerInto(ctx, log)
+	log.V(2).Info("Reconciling Pod")
+
+	// If this is the "leader" pod (completion index 0) and it has been scheduled,
+	// iterate through all follower pods, assigning them to the same topology as the leader.
+	if isLeaderPod(&pod) {
+		log.Info(fmt.Sprintf("%q is a leader pod", pod.Name))
+
+		if pod.Spec.NodeName == "" {
+			log.Info("leader has not been scheduled")
+			return ctrl.Result{}, nil
+		}
+		log.Info("assigning job pods to topology")
+		if err := r.assignJobPodsToTopology(ctx, &pod, topologyKey); err != nil {
+			log.Error(err, "assigning job pods to topology")
+			return ctrl.Result{}, err
+		}
+	} else {
+		log.Info(fmt.Sprintf("%q is a follower pod", pod.Name))
+
+		// If this is not a leader pod, check if the leader pod has been scheduled, and if so,
+		// assign this pod to the same topology.
+		log.Info("getting leader pod for follower")
+		leaderPod, err := r.leaderPodForFollower(ctx, &pod)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+
+		// If leader pod is not scheduled yet, do nothing and return.
+		if leaderPod.Spec.NodeName == "" {
```

Comment:
leaderPod could be nil, in which case we also need to return. I think this line is what caused the crash